File: services/eval/src/generator/rng.ts
/**
 * Deterministic pseudo-random number generator.
 *
 * WHY mulberry32: small, fast, decent distribution, fully deterministic from a
 * 32-bit seed. Sufficient for shuffling a chunk list; we are not doing crypto.
 */
export class SeededRng {
  private state: number;

  constructor(seed: number) {
    this.state = seed >>> 0;
  }

  next(): number {
    this.state = (this.state + 0x6d2b79f5) >>> 0;
    let t = this.state;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  }

  /**
   * Fisher-Yates shuffle in place.
   */
  shuffle<T>(items: T[]): T[] {
    for (let i = items.length - 1; i > 0; i--) {
      const j = Math.floor(this.next() * (i + 1));
      [items[i], items[j]] = [items[j], items[i]];
    }
    return items;
  }
}

File: services/eval/src/rag_eval/generator/rng.py
UINT32_MASK = 0xFFFFFFFF


class SeededRng:
    """Deterministic pseudo-random number generator (mulberry32).

    Bit-exact port of the TypeScript implementation: same seed produces the
    same sequence, so golden dataset sampling stays reproducible across the
    language migration. Not cryptographically secure — not needed here.
    """

    def __init__(self, seed: int) -> None:
        self._state = seed & UINT32_MASK

    def next(self) -> float:
        """Return the next float in [0, 1)."""
        self._state = (self._state + 0x6D2B79F5) & UINT32_MASK
        t = self._state
        t = ((t ^ (t >> 15)) * (t | 1)) & UINT32_MASK
        t ^= (t + (((t ^ (t >> 7)) * (t | 61)) & UINT32_MASK)) & UINT32_MASK
        t &= UINT32_MASK
        return (t ^ (t >> 14)) / 4294967296

    def shuffle[T](self, items: list[T]) -> list[T]:
        """Fisher-Yates shuffle in place."""
        for i in range(len(items) - 1, 0, -1):
            j = int(self.next() * (i + 1))
            items[i], items[j] = items[j], items[i]
        return items
